### PR TITLE
feat(plugin-meeting): add event for server media inactivity

### PR DESCRIPTION
**Pull request by arun3528**
added feature to emit server media inactivity for reconnect

https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-161433

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 1]
```diff
@@ -51,13 +51,43 @@ SelfUtils.getSelves = (oldSelf, newSelf, deviceId) => {
...
   };
 };
 
+SelfUtils.isMediaInactive = (previous, current) => {
+  if (previous && previous.joinedWith && previous.joinedWith.mediaSessions &&
+     current && current.joinedWith && current.joinedWith.mediaSessions) {
+    const previousMediaStatus = SelfUtils.getMediaStatus(previous.joinedWith.mediaSessions);
+    const currentMediaStatus = SelfUtils.getMediaStatus(current.joinedWith.mediaSessions);
+
+    console.log('Media Stream previous', previousMediaStatus);
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:69]: Can you remove these console logs?

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 2]
```diff
@@ -51,13 +51,43 @@ SelfUtils.getSelves = (oldSelf, newSelf, deviceId) => {
...
+    console.log('Media Stream current', currentMediaStatus);
+
+
+    if (previousMediaStatus.audio.state !== 'inactive' && currentMediaStatus.audio.state === 'inactive' && currentMediaStatus.audio.direction !== 'inactive') {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:73]: Do we have a constant for `'inactive'`?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 3]
```diff
@@ -1056,6 +1056,13 @@ export default class Meeting extends StatelessWebexPlugin {
         });
       }
     });
+
+    this.locusInfo.on(LOCUSINFO.EVENTS.MEDIA_INACTIVITY, (mediaStatus) => {
+      this.reconnect();
+      if (mediaStatus) {
+        // mediaStatus
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1063]: What is this?

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/locus-info/selfConstant.js [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+export const self = {
+  deviceUrl: 'https://wdm-a.wbx2.com/wdm/api/v1/devices/20eabde3-4254-48da-9a24',
+  alertHint: {
+    action: 'NONE'
+  },
+  alertType: {
+    action: 'NONE'
+  },
+  callbackInfo: {
+    callbackAddress: 'arungane@alpha.webex.com',
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/test/unit/spec/locus-info/selfConstant.js:11]: Can we remove some of the PII in here to generic?

File: packages/node_modules/samples/browser-plugin-meetings/app.js [hunk 5]
```diff
@@ -37,6 +37,18 @@ function initWebex() {
   authStatusElm.innerHTML = 'initializing...';
 
   webex = window.webex = Webex.init({
+    config: {
+      logger: {
+        level: 'debug'
+      },
+      meetings: {
+        deviceType: 'WEB',
```
Comment (adamweeks) [inline packages/node_modules/samples/browser-plugin-meetings/app.js:45]: We shouldn't be configuring device type

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 3 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1063]: in future if we want to indicate the user that you are not sending media as expected and call might reconnect or disconnect

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 6]
```diff
@@ -51,13 +52,39 @@ SelfUtils.getSelves = (oldSelf, newSelf, deviceId) => {
...
+  if (previous && previous.joinedWith && previous.joinedWith.mediaSessions &&
+     current && current.joinedWith && current.joinedWith.mediaSessions) {
+    const previousMediaStatus = SelfUtils.getMediaStatus(previous.joinedWith.mediaSessions);
+    const currentMediaStatus = SelfUtils.getMediaStatus(current.joinedWith.mediaSessions);
+
+    if (previousMediaStatus.audio.state !== MEDIA_STATE.inactive && currentMediaStatus.audio.state === MEDIA_STATE.inactive && currentMediaStatus.audio.direction !== MEDIA_STATE.inactive) {
+      return true;
+    }
+
+    if (previousMediaStatus.audio.state !== MEDIA_STATE.inactive && currentMediaStatus.audio.state === MEDIA_STATE.inactive && currentMediaStatus.audio.direction !== MEDIA_STATE.inactive) {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:74]: Why is this line repeated three times?

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 7]
```diff
@@ -194,6 +221,20 @@ SelfUtils.getSipUrl = (partner, type, sipUri) => {
...
 };
 
+SelfUtils.getMediaStatus = (mediaSessions = []) => {
+  const mediaStatus = {
+    audio: {},
+    video: {},
+    share: {}
+  };
+
+  mediaStatus.audio = mediaSessions.find((media) => media.mediaType === 'audio' && media.mediaContent === 'main');
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:231]: Media type is available as a constant in `import {eventType, trigger, mediaType} from '../metrics/config';`.
Do we have a constant for `mediaContent`?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 8]
```diff
@@ -1056,6 +1056,12 @@ export default class Meeting extends StatelessWebexPlugin {
         });
       }
     });
+
+    this.locusInfo.on(LOCUSINFO.EVENTS.MEDIA_INACTIVITY, () => {
+      this.reconnect();
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1061]: Do we need metrics for when we receive this event?

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js (hunk 6 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:74]: my bad it a copy past mistake . It should be audio video and share

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 8 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1061]: Its better to have one not sure if edonus sends one to the call-analyzer

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js (hunk 7 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:231]: no will create one

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 8 above)
Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1061]: Ok, we can investigate if we need one, not needed in this PR though.

**Comment by arun3528** [comment]
@adamweeks  pushed the fix

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 9]
```diff
@@ -51,13 +56,39 @@ SelfUtils.getSelves = (oldSelf, newSelf, deviceId) => {
...
   };
 };
 
+SelfUtils.isMediaInactive = (previous, current) => {
+  if (previous && previous.joinedWith && previous.joinedWith.mediaSessions &&
+     current && current.joinedWith && current.joinedWith.mediaSessions) {
+    const previousMediaStatus = SelfUtils.getMediaStatus(previous.joinedWith.mediaSessions);
+    const currentMediaStatus = SelfUtils.getMediaStatus(current.joinedWith.mediaSessions);
+
+    if (previousMediaStatus.audio.state !== MEDIA_STATE.inactive && currentMediaStatus.audio.state === MEDIA_STATE.inactive && currentMediaStatus.audio.direction !== MEDIA_STATE.inactive) {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:74]: Sorry, one more question about this check. Why would direction not being inactive mean media is inactive? Does `getMediaStatus` not update direction?

`currentMediaStatus.audio.direction !== MEDIA_STATE.inactive`

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 10]
```diff
@@ -1056,6 +1056,12 @@ export default class Meeting extends StatelessWebexPlugin {
...
+      // IF needed we can send an event to the develper when media gets inactive
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1062]:
```suggestion
      // IF needed we can send an event to the developer when media gets inactive
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 11]
```diff
@@ -1056,6 +1056,12 @@ export default class Meeting extends StatelessWebexPlugin {
...
+      // but anyways we are going to reconnect at this state
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1063]: You can remove this line.

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js (hunk 9 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:74]: If someone had not request for media like video to be sent and received intentionally. Then we should not reconnect as its a valid scenario 

So the direction and status is the one we comparer if its sendOnly , recvonly and sendRecv but actually the media was inactive then we trigger reconenct

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 12]
```diff
@@ -51,13 +56,39 @@ SelfUtils.getSelves = (oldSelf, newSelf, deviceId) => {
...
+
+    if (previousMediaStatus.share.state !== MEDIA_STATE.inactive && currentMediaStatus.share.state === MEDIA_STATE.inactive && currentMediaStatus.share.direction !== MEDIA_STATE.inactive) {
+      return true;
+    }
+
+    return false;
+  }
+
+  return false;
+};
```
Comment (taymoork2) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:90]: non-blocking formatting suggestion

```suggestion
SelfUtils.isMediaInactive = (previous, current) => {
  if (
    previous &&
    previous.joinedWith &&
    previous.joinedWith.mediaSessions &&
    current &&
    current.joinedWith &&
    current.joinedWith.mediaSessions
  ) {
    const previousMediaStatus = SelfUtils.getMediaStatus(
      previous.joinedWith.mediaSessions
    );
    const currentMediaStatus = SelfUtils.getMediaStatus(
      current.joinedWith.mediaSessions
    );

    if (
      previousMediaStatus.audio.state !== MEDIA_STATE.inactive &&
      currentMediaStatus.audio.state === MEDIA_STATE.inactive &&
      currentMediaStatus.audio.direction !== MEDIA_STATE.inactive
    ) {
      return true;
    }

    if (
      previousMediaStatus.video.state !== MEDIA_STATE.inactive &&
      currentMediaStatus.video.state === MEDIA_STATE.inactive &&
      currentMediaStatus.video.direction !== MEDIA_STATE.inactive
    ) {
      return true;
    }

    if (
      previousMediaStatus.share.state !== MEDIA_STATE.inactive &&
      currentMediaStatus.share.state === MEDIA_STATE.inactive &&
      currentMediaStatus.share.direction !== MEDIA_STATE.inactive
    ) {
      return true;
    }

    return false;
  }

  return false;
};
```

**Comment by adamweeks** [comment]
@arun3528 please split the samples changes into their own commits. We have the changelog being generated now from the commits.

`feat(samples): enable reconnection to meeting sample` or something similar.

**Comment by arun3528** [comment]
@adamweeks @taymoork2  splitted into two PR

**Comment by adamweeks** [comment]
LGTM!

**Pull request merged**